Register Hibernate @GenericGenerator strategy implementations for reflective access

@Sanne (as discussed on Zulip):
I have an entity using org.hibernate.id.UUIDGenerator:
```
public class AuthenticationRequest {

    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(
            name = "UUID",
            strategy = "org.hibernate.id.UUIDGenerator"
    )
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id;
```
This works fine in Java but when running in native mode I get:
```
Caused by: org.hibernate.MappingException: Could not interpret id generator strategy [org.hibernate.id.UUIDGenerator]
    at org.hibernate.id.factory.internal.DefaultIdentifierGeneratorFactory.getIdentifierGeneratorClass(DefaultIdentifierGeneratorFactory.java:144)
    at org.hibernate.mapping.SimpleValue.isIdentityColumn(SimpleValue.java:378)
    at org.hibernate.tool.schema.internal.StandardTableExporter.getSqlCreateStrings(StandardTableExporter.java:60)
    at org.hibernate.tool.schema.internal.StandardTableExporter.getSqlCreateStrings(StandardTableExporter.java:30)
    at org.hibernate.tool.schema.internal.SchemaCreatorImpl.createFromMetadata(SchemaCreatorImpl.java:316)
    at org.hibernate.tool.schema.internal.SchemaCreatorImpl.performCreation(SchemaCreatorImpl.java:166)
    at org.hibernate.tool.schema.internal.SchemaCreatorImpl.doCreation(SchemaCreatorImpl.java:135)
    at org.hibernate.tool.schema.internal.SchemaCreatorImpl.doCreation(SchemaCreatorImpl.java:121)
    at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:155)
```
Kabir Khan: Is there some trick I need to perform, or is it a bug?
Kabir Khan: BTW I am planning on ripping out this entity so it isn't a huge deal
Sanne Grinovero: @Kabir Khan ah! thanks that's a useful bugreport :) Yes we probably need to register those strategy implementations for reflective access, didn't think of that.
Kabir Khan: Cool

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/registry/PreconfiguredServiceRegistryBuilder.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/registry/PreconfiguredServiceRegistryBuilder.java
@@ -28,7 +28,6 @@ import org.hibernate.boot.registry.StandardServiceInitiator;
 import org.hibernate.boot.registry.internal.BootstrapServiceRegistryImpl;
 import org.hibernate.boot.registry.internal.StandardServiceRegistryImpl;
 import org.hibernate.boot.registry.selector.internal.StrategySelectorImpl;
-import org.hibernate.cache.internal.RegionFactoryInitiator;
 import org.hibernate.engine.config.internal.ConfigurationServiceInitiator;
 import org.hibernate.engine.jdbc.batch.internal.BatchBuilderInitiator;
 import org.hibernate.engine.jdbc.connections.internal.ConnectionProviderInitiator;
@@ -41,7 +40,6 @@ import org.hibernate.engine.transaction.jta.platform.internal.JtaPlatformInitiat
 import org.hibernate.engine.transaction.jta.platform.spi.JtaPlatform;
 import org.hibernate.event.internal.EntityCopyObserverFactoryInitiator;
 import org.hibernate.hql.internal.QueryTranslatorFactoryInitiator;
-import org.hibernate.id.factory.internal.MutableIdentifierGeneratorFactoryInitiator;
 import org.hibernate.integrator.spi.Integrator;
 import org.hibernate.internal.EntityManagerMessageLogger;
 import org.hibernate.internal.util.config.ConfigurationHelper;
@@ -58,7 +56,12 @@ import org.hibernate.tool.hbm2ddl.ImportSqlCommandExtractorInitiator;
 import org.hibernate.tool.schema.internal.SchemaManagementToolInitiator;
 
 import io.quarkus.hibernate.orm.runtime.recording.RecordedState;
-import io.quarkus.hibernate.orm.runtime.service.*;
+import io.quarkus.hibernate.orm.runtime.service.CfgXmlAccessServiceInitiatorQuarkus;
+import io.quarkus.hibernate.orm.runtime.service.DisabledJMXInitiator;
+import io.quarkus.hibernate.orm.runtime.service.FlatClassLoaderService;
+import io.quarkus.hibernate.orm.runtime.service.QuarkusJdbcEnvironmentInitiator;
+import io.quarkus.hibernate.orm.runtime.service.QuarkusJtaPlatformResolver;
+import io.quarkus.hibernate.orm.runtime.service.QuarkusRegionFactoryInitiator;
 
 /**
  * Helps to instantiate a ServiceRegistryBuilder from a previous state. This
@@ -200,7 +203,9 @@ public class PreconfiguredServiceRegistryBuilder {
         serviceInitiators.add(RefCursorSupportInitiator.INSTANCE);
 
         serviceInitiators.add(QueryTranslatorFactoryInitiator.INSTANCE);
-        serviceInitiators.add(MutableIdentifierGeneratorFactoryInitiator.INSTANCE);
+
+        // Disabled: IdentifierGenerators are no longer initiated after Metadata was generated.
+        // serviceInitiators.add(MutableIdentifierGeneratorFactoryInitiator.INSTANCE);
 
         // Replaces JtaPlatformResolverInitiator.INSTANCE );
         serviceInitiators.add(new QuarkusJtaPlatformResolver(rs.getJtaPlatform()));

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/recording/RecordableBootstrap.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/recording/RecordableBootstrap.java
@@ -67,6 +67,7 @@ import org.hibernate.tool.schema.internal.SchemaManagementToolInitiator;
 import io.quarkus.hibernate.orm.runtime.boot.QuarkusEnvironment;
 import io.quarkus.hibernate.orm.runtime.service.DialectFactoryInitiator;
 import io.quarkus.hibernate.orm.runtime.service.DisabledJMXInitiator;
+import io.quarkus.hibernate.orm.runtime.service.QuarkusMutableIdentifierGeneratorFactoryInitiator;
 import io.quarkus.hibernate.orm.runtime.service.QuarkusRegionFactoryInitiator;
 
 /**
@@ -136,7 +137,9 @@ public final class RecordableBootstrap extends StandardServiceRegistryBuilder {
         serviceInitiators.add(RefCursorSupportInitiator.INSTANCE);
 
         serviceInitiators.add(QueryTranslatorFactoryInitiator.INSTANCE);
-        serviceInitiators.add(MutableIdentifierGeneratorFactoryInitiator.INSTANCE);
+
+        // Custom one! Also, this one has state so can't use the singleton.
+        serviceInitiators.add(new QuarkusMutableIdentifierGeneratorFactoryInitiator());// MutableIdentifierGeneratorFactoryInitiator.INSTANCE);
 
         serviceInitiators.add(JtaPlatformResolverInitiator.INSTANCE);
         serviceInitiators.add(JtaPlatformInitiator.INSTANCE);

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/service/QuarkusMutableIdentifierGeneratorFactory.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/service/QuarkusMutableIdentifierGeneratorFactory.java
@@ -1,0 +1,82 @@
+package io.quarkus.hibernate.orm.runtime.service;
+
+import java.io.Serializable;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.hibernate.dialect.Dialect;
+import org.hibernate.id.IdentifierGenerator;
+import org.hibernate.id.factory.internal.DefaultIdentifierGeneratorFactory;
+import org.hibernate.id.factory.spi.MutableIdentifierGeneratorFactory;
+import org.hibernate.service.spi.ServiceRegistryAwareService;
+import org.hibernate.service.spi.ServiceRegistryImplementor;
+import org.hibernate.type.Type;
+
+/**
+ * Wraps the default DefaultIdentifierGeneratorFactory so to make sure we store the Class references
+ * of any IdentifierGenerator which is accessed during the build of the Metadata.
+ *
+ * This is not to register them for reflection access: all reflective instantiation is performed
+ * during the build of the Metadata and is therefore safe even in native mode;
+ * however we still need the Class instances as some runtime operations will need these, and
+ * will look them up by either fully qualified name (and then reflection) or strategy name.
+ *
+ * Since all IdentifierGenerator types used by a model are accessed during the Metadata creation,
+ * just watching for these will provide the full list of Class instances we need to keep.
+ */
+public final class QuarkusMutableIdentifierGeneratorFactory
+        implements MutableIdentifierGeneratorFactory, Serializable, ServiceRegistryAwareService {
+
+    private final DefaultIdentifierGeneratorFactory original = new DefaultIdentifierGeneratorFactory();
+    private final ConcurrentHashMap<String, Class<? extends IdentifierGenerator>> typeCache = new ConcurrentHashMap<>();
+
+    @Override
+    public void register(final String strategy, final Class generatorClass) {
+        original.register(strategy, generatorClass);
+        storeCache(strategy, generatorClass);
+    }
+
+    @Override
+    public Dialect getDialect() {
+        return original.getDialect();
+    }
+
+    @Override
+    public void setDialect(final Dialect dialect) {
+        //currently a no-op anyway..?
+        original.setDialect(dialect);
+    }
+
+    @Override
+    public IdentifierGenerator createIdentifierGenerator(final String strategy, final Type type, final Properties config) {
+        final IdentifierGenerator identifierGenerator = original.createIdentifierGenerator(strategy, type, config);
+        storeCache(strategy, identifierGenerator.getClass());
+        return identifierGenerator;
+    }
+
+    private void storeCache(final String strategy, final Class<? extends IdentifierGenerator> generatorClass) {
+        if (strategy == null || generatorClass == null)
+            return;
+        final String className = generatorClass.getName();
+        //Store for access both via short and long names:
+        typeCache.put(strategy, generatorClass);
+        if (!className.equals(strategy)) {
+            typeCache.put(className, generatorClass);
+        }
+    }
+
+    @Override
+    public Class getIdentifierGeneratorClass(final String strategy) {
+        Class<? extends IdentifierGenerator> aClass = typeCache.get(strategy);
+        if (aClass != null)
+            return aClass;
+        aClass = original.getIdentifierGeneratorClass(strategy);
+        storeCache(strategy, aClass);
+        return aClass;
+    }
+
+    @Override
+    public void injectServices(final ServiceRegistryImplementor serviceRegistry) {
+        original.injectServices(serviceRegistry);
+    }
+}

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/service/QuarkusMutableIdentifierGeneratorFactoryInitiator.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/service/QuarkusMutableIdentifierGeneratorFactoryInitiator.java
@@ -1,0 +1,33 @@
+package io.quarkus.hibernate.orm.runtime.service;
+
+import java.util.Map;
+
+import org.hibernate.boot.registry.StandardServiceInitiator;
+import org.hibernate.id.factory.internal.MutableIdentifierGeneratorFactoryInitiator;
+import org.hibernate.id.factory.spi.MutableIdentifierGeneratorFactory;
+import org.hibernate.service.spi.ServiceRegistryImplementor;
+
+/**
+ * Needs to mimick MutableIdentifierGeneratorFactoryInitiator, but allows us to capture
+ * which Identifier strategies are being used, so that we can keep a reference to the classed
+ * needed at runtime.
+ *
+ * @see MutableIdentifierGeneratorFactoryInitiator
+ */
+public final class QuarkusMutableIdentifierGeneratorFactoryInitiator
+        implements StandardServiceInitiator<MutableIdentifierGeneratorFactory> {
+
+    private final MutableIdentifierGeneratorFactory sfScopedSingleton = new QuarkusMutableIdentifierGeneratorFactory();
+
+    @Override
+    public MutableIdentifierGeneratorFactory initiateService(final Map configurationValues,
+            final ServiceRegistryImplementor registry) {
+        return sfScopedSingleton;
+    }
+
+    @Override
+    public Class<MutableIdentifierGeneratorFactory> getServiceInitiated() {
+        return MutableIdentifierGeneratorFactory.class;
+    }
+
+}

--- a/integration-tests/jpa-h2/src/main/java/io/quarkus/example/jpah2/Person.java
+++ b/integration-tests/jpa-h2/src/main/java/io/quarkus/example/jpah2/Person.java
@@ -1,38 +1,42 @@
 package io.quarkus.example.jpah2;
 
+import java.util.UUID;
+
 import javax.persistence.CascadeType;
 import javax.persistence.Entity;
 import javax.persistence.FetchType;
 import javax.persistence.GeneratedValue;
-import javax.persistence.GenerationType;
 import javax.persistence.Id;
 import javax.persistence.ManyToOne;
 import javax.persistence.NamedQuery;
+
+import org.hibernate.annotations.GenericGenerator;
 
 @Entity
 @NamedQuery(name = "get_person_by_name", query = "select p from Person p where name = :name")
 public class Person {
 
-    private long id;
+    private UUID id;
     private String name;
     private SequencedAddress address;
 
     public Person() {
     }
 
-    public Person(long id, String name, SequencedAddress address) {
+    public Person(UUID id, String name, SequencedAddress address) {
         this.id = id;
         this.name = name;
         this.address = address;
     }
 
     @Id
-    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "personSeq")
-    public long getId() {
+    @GeneratedValue(generator = "UUID")
+    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
+    public UUID getId() {
         return id;
     }
 
-    public void setId(long id) {
+    public void setId(UUID id) {
         this.id = id;
     }
 

